### PR TITLE
Issue #5779: fixing no violation on parenthesis in if statement betwe…

### LINKED ISSUE #5779: UnnecessaryParentheses: no violation on parenthesis in if statement between || and &&
Identified at https://github.com/checkstyle/checkstyle/pull/5778#discussion_r186302647 ,

````
$ cat TestClass.java
public class TestClass {
    void method(String sectionName) {
        if ("Content".equals(sectionName) || "Overview".equals(sectionName)
                // suppression list
                || (!"AbbreviationAsWordInName".equals(sectionName)
                        && !"AbstractClassName".equals(sectionName)
                        )) {
             continue;
         }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="UnnecessaryParentheses" />
    </module>
</module>

$ java -jar checkstyle-8.10-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

Every other CI complains about parenthesis at line 5 except our check. There should be a violation there.

### PR DESCRIPTION
**Pull request by Vyom-Yadav**
Continuing #9308 
Resolves #5779: fixing no violation on parenthesis in if statement between || and &&
Diff Regression projects: https://gist.githubusercontent.com/Vyom-Yadav/91807e6244cff60698d9e33e32ba2d51/raw/85355576c05c80e8b08752a62ba3a4076909d266/projects-to-test-on.properties
Diff Regression config: https://gist.githubusercontent.com/Vyom-Yadav/98dceb63a79f4833e85fff9b2e1464a6/raw/7a57a856b6657d8606276270a6ca3276d102ef65/my_checks.xml

### CODE REVIEW

**Comment by Vyom-Yadav** [comment]
I couldn't run `./.ci/wercker.sh no-error-orekit` on local, the build was successful for Hipparchus but it failed on orekit-
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (default-cli) on project orekit: Execution default-cli of goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check failed: Plugin org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1 or one of its dependencies could not be resolved: Failure to find com.puppycrawl.tools:checkstyle:jar:8.41-SNAPSHOT in https://repository.apache.org/snapshots was cached in the local repository, resolution will not be reattempted until the update interval of apache.snapshots has elapsed or updates are forced
```
[The whole log](https://gist.githubusercontent.com/Vyom-Yadav/411a2e93c49f9d7ae8411b974031e542/raw/41b2e47444b366b86d4fa97fc902de64420ecec5/log2.txt)

**Comment by Vyom-Yadav** [comment]
Github, generate report

**Comment by Vyom-Yadav** [comment]
> I couldn't run `./.ci/wercker.sh no-error-orekit` on local, the build was successful for Hipparchus but it failed on orekit-
> 
> ```
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (default-cli) on project orekit: Execution default-cli of goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check failed: Plugin org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1 or one of its dependencies could not be resolved: Failure to find com.puppycrawl.tools:checkstyle:jar:8.41-SNAPSHOT in https://repository.apache.org/snapshots was cached in the local repository, resolution will not be reattempted until the update interval of apache.snapshots has elapsed or updates are forced
> ```
> 
> [The whole log](https://gist.githubusercontent.com/Vyom-Yadav/411a2e93c49f9d7ae8411b974031e542/raw/41b2e47444b366b86d4fa97fc902de64420ecec5/log2.txt)

@strkkk @rnveach can you take a look?

**Comment by rnveach** [comment]
> ArtifactNotFoundException: Failure to find com.puppycrawl.tools:checkstyle:jar:8.41-SNAPSHOT

Make sure your branch is rebased on the latest master. A release just occurred and failure to find a snapshot version means PR version of checkstyle was not installed. SNAPSHOTs don't go to maven servers.

**Comment by rnveach** [comment]
> ArtifactNotFoundException: Failure to find com.puppycrawl.tools:checkstyle:jar:8.41-SNAPSHOT

@Vyom-Yadav Sorry, it seems wercker is a little special.
You must run checkstyle install first.
https://github.com/checkstyle/checkstyle/blob/db7215dc8b80fa99c27250c0548e9581f0f40208/wercker.yml#L60

@romani I think we should relook at the scripts if this action is required to run the command locally. I would expect the script to do this and not require manual intervention.

**Comment by Vyom-Yadav** [comment]
@rnveach I sent a PR to the Orekit project https://gitlab.orekit.org/orekit/orekit/-/merge_requests/143

**Comment by Vyom-Yadav** [comment]
They merged the changes but still, the wrecker is failing, I checked in my local, it showing the same set of violations even though they have been fixed.

**Comment by Vyom-Yadav** [comment]
@romani @rnveach Ready for review.

**Comment by Vyom-Yadav** [comment]
Why is the spell check failing

**Comment by Vyom-Yadav** [comment]
Why is spell check failing 😕

**Comment by Vyom-Yadav** [comment]
@romani @rnveach Ready for review.

**Comment by strkkk** [comment]
@Vyom-Yadav sonar is not happy
https://sonarcloud.io/project/issues?id=org.checkstyle%3Acheckstyle&pullRequest=9334&resolved=false&severities=CRITICAL&sinceLeakPeriod=true

**Comment by Vyom-Yadav** [comment]
> @Vyom-Yadav sonar is not happy
> https://sonarcloud.io/project/issues?id=org.checkstyle%3Acheckstyle&pullRequest=9334&resolved=false&severities=CRITICAL&sinceLeakPeriod=true

@strkkk I fixed it, ready for review.

**Comment by Vyom-Yadav** [comment]
@romani @strkkk ping.

File: .ci/wercker.sh [hunk 1]
```diff
@@ -91,7 +91,7 @@ no-error-orekit)
   # no CI is enforced in project, so to make our build stable we should
   # checkout to latest release/development (annotated tag or hash) or sha that have fix we need
   # git checkout $(git describe --abbrev=0 --tags)
-  git checkout "54f5f2eb410ec42e5410ec8e5c415ff9182d3235"
+  git checkout "76760bfbfb847e227490cd5d3662fca087f1a324"
```
Comment (strkkk) [inline .ci/wercker.sh:94]: please add double quotes in places where spellchecker reports violation, e.g. as""db""123 etc.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 2]
```diff
@@ -437,6 +516,51 @@ private static boolean isExprSurrounded(DetailAST ast) {
...
+     *         parentheses.
+     */
+    private static boolean checkAroundOperators(DetailAST ast) {
+        final int type = ast.getType();
+        final DetailAST parent = ast.getParent();
+        return (isInTokenList(type, CONDITIONALS)
+                    || isInTokenList(type, RELATIONAL)
+                        || isInTokenList(type, UNARY_AND_POSTFIX))
+                && (isInTokenList(parent.getType(), CONDITIONALS)
+                    || isInTokenList(parent.getType(), RELATIONAL))
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:560]: Please use TokenUtil.isOfType and remove isInTokenList method

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 3]
```diff
@@ -437,6 +516,51 @@ private static boolean isExprSurrounded(DetailAST ast) {
...
+     * @param ast the {@code DetailAST} to check if it is surrounded by
+     *        parentheses.
+     * @return {@code true} if the expression is surrounded by
+     *         parentheses.
+     */
+    private static boolean checkAroundOperators(DetailAST ast) {
+        final int type = ast.getType();
+        final DetailAST parent = ast.getParent();
+        return (isInTokenList(type, CONDITIONALS)
+                    || isInTokenList(type, RELATIONAL)
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:557]: I think it is better to merge CONDITIONALS and RELATIONAL lists into one and simplify this checks. Same for lines below

**Review by strkkk — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 2 above)
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: Done, shall I completely remove the usage of isInTokenList from everywhere in check and use TokenUtil.isOfType?

File: .ci/wercker.sh (hunk 1 above)
Comment (Vyom-Yadav) [reply .ci/wercker.sh]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 3 above)
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: done.

File: .ci/wercker.sh (hunk 1 above)
Comment (Vyom-Yadav) [reply .ci/wercker.sh]: I tried using-

- `"76760""bfbfb""847e227490cd5d3662fca""087f1a324"`

- `"76760""bfbfb""847e227490cd5d3662""fca""087f1a324"`

- `"76760""bfbfb847e2""27490cd5d""3662fca""087f1a324"`
But none of them worked, all showed the same error, can you point out where am I going wrong, spellchecker complains about the words-

- `bfbfb`
- `fca`

Comment (strkkk) [reply .ci/wercker.sh]: you need to make it like "76760bf""bf""b847e227490cd5d3662f""ca087f1a324"

I remember someone told me that spellchecker will not complain if there is only 2 letters, so you need double quotes to split this letter sequences

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 2 above)
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: you can do it in second commit with `minor:` prefix

**Comment by strkkk** [comment]
@Vyom-Yadav please fix CI

File: .ci/wercker.sh (hunk 1 above)
Comment (Vyom-Yadav) [reply .ci/wercker.sh]: done.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 2 above)
Comment (Vyom-Yadav) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: done.

**Review by strkkk — APPROVED** [review]

**Review by pbludov — APPROVED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+        if ((Integer.valueOf(x) instanceof Integer) // violation, unnecessary parenthesis
+                || Integer.valueOf(y) instanceof Integer) { // ok
+            return;
+        }
+        if (x == ((y<z) ? y : z) &&
+            ((x>y && y>z) || (!(x<z) && y>z))) { // violation, unnecessary parenthesis
+            return;
+        }
+        if((x >= 0 && y <= 9) || (z >= 5 && y <= 5) || (z >= 3 && x <= 7)) { // violation,
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:53]: Split this into separate lines so it is clear where the violation is.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        if (((x == y) || (x != z))) { // violation, unnecessary parenthesis
+            return;
+        }
+
+        if ((Integer.valueOf(x) instanceof Integer) // violation, unnecessary parenthesis
+                || Integer.valueOf(y) instanceof Integer) { // ok
+            return;
+        }
+        if (x == ((y<z) ? y : z) &&
+            ((x>y && y>z) || (!(x<z) && y>z))) { // violation, unnecessary parenthesis
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:50]: Same, or atleast provide some guidance if all of them are violations.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        if ((x < y) && (x > z)) { // violation, unnecessary parenthesis
+            return;
+        }
+
+        if (((x < y) && (x > z))) { // violation, unnecessary parenthesis
+            return;
+        }
+
+        if (!(x <= y) || (x >= z)) { // violation, unnecessary parenthesis
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:33]: I would say the same with the rest. Its not clear if the violation is on the left or right hand side. I remember this was a question for me previously.

**Review by rnveach — CHANGES_REQUESTED** [review]
none of the test cases show cases where no violation would be printed. Alot of them only go to 2 conditions.

What about 3 and using various `||`s and `&&`s.

**Review by rnveach — CHANGES_REQUESTED** [review]
Please also update description for check at https://checkstyle.org/config_coding.html#UnnecessaryParentheses_Description and examples at https://checkstyle.org/config_coding.html#UnnecessaryParentheses_Examples .

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java (hunk 4 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java]: done.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java (hunk 5 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java (hunk 6 above)
Comment (Vyom-Yadav) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java]: done

**Comment by Vyom-Yadav** [comment]
> Please also update description for check at https://checkstyle.org/config_coding.html#UnnecessaryParentheses_Description and examples at https://checkstyle.org/config_coding.html#UnnecessaryParentheses_Examples .

I updated those sections also made the test cases more clear.

**Comment by Vyom-Yadav** [comment]
Github, generate web site

**Comment by rnveach** [comment]
> none of the test cases show cases where no violation would be printed. Alot of them only go to 2 conditions.
> What about 3 and using various ||s and &&s.

I am not seeing a comment if these were done.

**Comment by rnveach** [comment]
Example:
````
if (cond1 AND cond2 AND cond3 OR cond4 AND cond5 AND cond6)
````
is the same as:
````
if ((cond1 AND cond2 AND cond3) OR (cond4 AND cond5 AND cond 6))
````
which makes those parenthesis redundant.

But if I write it like:
````
if (cond1 AND cond2 AND (cond3 OR cond4) AND cond5 AND cond6)
````
Suddenly they aren't redundant and the parenthesis have a purpose to change the order of operations. There shouldn't be a violation here.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                 || (z >= 3 && x <= 7)) { // violation, unnecessary parenthesis
+            return;
+        }
+        if(x>= 0 && (x<=8 || y<=11) && y>=8) { // ok
+            return;
+        }
+        if((y>=11 && x<=5)            // violation, unnecessary parenthesis
+                || (x<=12 && y>=8)) { // violation, unnecessary parenthesis
+            return;
+        }
```
Comment (Vyom-Yadav) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIfStatement.java:72]: @rnveach Done.

**Review by romani — APPROVED** [review]
I see tests that on items that @rnveach  requested, ok to merge.

**Pull request merged**